docker/service.go

```diff
@@ -186,15 +200,59 @@ func (s *Service) up(imageName string, create bool) error {
 	}
 
 	return s.eachContainer(func(c *Container) error {
-		if outOfSync, err := c.OutOfSync(); err != nil {
-			return err
-		} else if outOfSync {
-			logrus.Warnf("%s needs rebuilding", s.Name())
+		if create {
+			if err := s.rebuildIfNeeded(c); err != nil {
+				return err
+			}
 		}
+
 		return c.Up(imageName)
 	})
 }
 
+func (s *Service) rebuildIfNeeded(c *Container) error {
+	imageName, err := s.build()
+	if err != nil {
+		return err
+	}
+
+	if s.context.NoRecreate {
+		return nil
+	}
+
+	outOfSync, err := c.OutOfSync()
+	if err != nil {
+		return err
+	}
+
+	containerInfo, err := c.findInfo()
+	if containerInfo == nil || err != nil {
+		return err
+	}
+	name := containerInfo.Name[1:]
+
+	origRebuildLabel := containerInfo.Config.Labels[REBUILD.Str()]
+	newRebuildLabel := s.Config().Labels.MapParts()[REBUILD.Str()]
+	rebuildLabelChanged := newRebuildLabel != origRebuildLabel
+	logrus.WithFields(logrus.Fields{
+		"origRebuildLabel":    origRebuildLabel,
+		"newRebuildLabel":     newRebuildLabel,
+		"outOfSync":           outOfSync,
+		"ForceRecreate":       s.context.ForceRecreate,
+		"rebuildLabelChanged": rebuildLabelChanged}).Debug("Rebuild values")
+
+	if s.context.ForceRecreate || origRebuildLabel == "always" || rebuildLabelChanged || origRebuildLabel != "false" && outOfSync {
+		logrus.Infof("Rebuilding %s", name)
+		if _, err := c.Rebuild(imageName); err != nil {
+			return err
+		}
```

Comment:
@dnephin Actually let me explain this logic a bit more.  There are two values you can put for the label REBUILD=always or REBUILD=false.  The context here is specially the "console" container in RancherOS.  The way RancherOS works is that the when the OS boots we essentially do "up".  We have three modes for the console: ephemeral, persistant, "default".

1. ephemeral: An every up we always recreate, the is the equivalent of `--force-recreate`.
1. persistant: Never recreate the container regardless if the compose file changes.  Basically the equivalent of `--no-recreate`
1. default: just normal logic

The logic gets confusing when you recreate container with different values of this REBUILD tag.  For example if REBUILD=false and then you edit the compose file to REBUILD=always or vice versa.  I'll tell you we tested this extensively.

We really need this logic, so if you don't want it in libcompose we will just have to find away of pulling it out.  The key thing here is that we need to essentially set `--force-recreate/--no-recreate` on a specific service in the project and ignore the which mode is set for the current up.

Your call.  I see two paths forward.  We can keep the feature but change the label and terminology to fall more inline with docker-compose, so RECREATE=no or RECREATE=force.  Or we yank it out and just do it in RancherOS.  I think it's a useful feature.